lsl-viewer.py

```diff
@@ -104,27 +104,38 @@ def __init__(self, stream, fig, axes,  window, scale, dejitter=True):
 
         self.bf, self.af = butter(4, np.array([1, 40])/(self.sfreq/2.),
                                   'bandpass')
+        zi = lfilter_zi(self.bf, self.af)
+        self.filt_state = np.tile(zi, (self.n_chan, 1)).transpose()
+        self.filt_buffer = np.zeros((0, self.n_chan))
 
     def update_plot(self):
         k = 0
         while self.started:
             samples, timestamps = self.inlet.pull_chunk(timeout=1.0,
                                                         max_samples=12)
             if timestamps:
-                self.data = np.vstack([self.data, samples])
                 if self.dejitter:
                     timestamps = np.float64(np.arange(len(timestamps)))
                     timestamps /= self.sfreq
                     timestamps += self.times[-1] + 1./self.sfreq
                 self.times = np.concatenate([self.times, timestamps])
-
                 self.n_samples = int(self.sfreq * self.window)
-                self.data = self.data[-self.n_samples:]
                 self.times = self.times[-self.n_samples:]
+                self.data = np.vstack([self.data, samples])
+                self.data = self.data[-self.n_samples:]
+
+                if self.filt:
+                    self.filt_buffer = np.vstack([self.filt_buffer, samples])
                 k += 1
                 if k == self.display_every:
                     if self.filt:
-                        data_f = filtfilt(self.bf, self.af, self.data, axis=0)
+                        filt_samples, self.filt_state = lfilter(self.bf, self.af,
+                                                                    self.filt_buffer, axis=0,
+                                                                    zi=self.filt_state)
+                        data_f = np.vstack([self.data[:-len(self.filt_buffer)], filt_samples])
+                        self.filt_buffer = np.zeros((0, self.n_chan))
+                        self.data = data_f
```

Comment:
i would actually prefer to save data_f as a second buffer, so that you can switch between filtered and unfiltered data (keystroke 'd')